### DISCUSSION TITLE
Are sharable config presets supported for Azure DevOps?

### DISCUSSION THREAD

**Discussion by tomkerkhove** [renovatebot/renovate#10671] [Request Help] [answered]
I'm looking at [local presets](https://docs.renovatebot.com/config-presets/#local-presets) to have a central Renovate template for all our Azure DevOps Git repos and per the docs I'm using the following:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "<project-name>/<repo-name>"
  ]
}
```

The `renovate.json` is in the root of both repos, but unfortunately it is now using the centralized one. It made me wonder if this is supported in Azure DevOps but based on the docs of Local Preset it sounds like it but wanted to double check.

**Comment by viceice** [comment, marked as answer]
Sure, local presets are supported by all renovate platforms.

**Reply by tomkerkhove** [reply]
Does the configuration look correctly then because it's ignoring the shared template.

**Reply by viceice** [reply]
`<project-name>/<repo-name>` this repo must be readable by renovate, and the file name must be `default.json` in that repo root.

**Reply by tomkerkhove** [reply]
That did the trick, however, according to the docs `renovate.json` should have worked as well 🤔

Anyway, thank you!

**Reply by viceice** [reply]
yes, `renovate.json` is deprecated. you should prefer `default.json`, as the other is a fallback only.

@HonkingGoose we should add a docs hint, that `renovate.json` is deprecated and `default.json` should be used.

**Reply by tomkerkhove** [reply]
We might want to update the docs then.

**Reply by HonkingGoose** [reply]
> yes, `renovate.json` is deprecated. you should prefer `default.json`, as the other is a fallback only.
> 
> @HonkingGoose we should add a docs hint, that `renovate.json` is deprecated and `default.json` should be used.

Huh? I still use a `renovate.json` in all my repositories... Should I be using a `default.json` file from now on then? I don't use any shared presets, so I do not have a `honkinggoose/.github` repo with a `default.json` file in it. 😉 

---

Also a quick search through the docs yields 66 results for the string `renovate.json` in all `*.md` files, so I need to be sure where/what I need to change. 😄

**Reply by viceice** [reply]
this file (one example).

https://github.com/renovatebot/renovate/blob/5f325748b1ea406764b74354fa121d259982d901/docs/usage/config-presets.md#L190-L190

the `default.json` is ment for preset files only! We deprecate the `renovate.json` for presets, because it can collide with the repo config `renovate.json`

**Comment by kipusoep** [comment]
Sorry for bumping this old discussion, but I'm having a hard time getting this to work with Azure DevOps.
My config in the repo:
```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [ "local>Project Name/Infrastructure" ]
}
```

I have a `default.json` in the main branch of the Infrastructure repository.

When I run the pipeline, I get:
```
TRACE: resolveConfigPresets (repository=Project Name/TargetRepo)
       "config": {
         "$schema": "[https://docs.renovatebot.com/renovate-schema.json",](https://docs.renovatebot.com/renovate-schema.json%22,)
         "extends": ["local>Project Name/Infrastructure"]
       },
       "existingPresets": []
TRACE: Resolving preset "local>Project Name/Infrastructure" (repository=Project Name/TargetRepo)

TRACE: getPreset(local>Project Name/Infrastructure) (repository=Project Name/TargetRepo)

TRACE: getRepoByName(Project Name/Infrastructure) (repository=Project Name/TargetRepo)

TRACE: getProjectAndRepo(Project Name/Infrastructure) (repository=Project Name/TargetRepo)

DEBUG: Failed to retrieve default.json from repo Project Name/Infrastructure (repository=Project Name/TargetRepo)
       "err": {
         "message": "Cannot read property 'id' of null",
         "stack": "TypeError: Cannot read property 'id' of null\n    at Proxy.getRawFile (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/modules/platform/azure/index.js:59:23)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async fetchJSONFile (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/config/presets/local/common.js:11:15)\n    at async fetchPreset (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/config/presets/util.js:20:27)\n    at async getPreset (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/config/presets/index.js:178:24)\n    at async Object.resolveConfigPresets (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/config/presets/index.js:243:37)\n    at async mergeRenovateConfig (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/workers/repository/init/merge.js:192:61)\n    at async getRepoConfig (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/workers/repository/init/config.js:15:14)\n    at async initRepo (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/workers/repository/init/index.js:26:14)\n    at async Object.renovateRepository (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/workers/repository/index.js:37:18)\n    at async Object.start (/home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/workers/global/index.js:116:13)\n    at async /home/vsts/.npm/_npx/1667/lib/node_modules/renovate/dist/renovate.js:15:24"
       },
       "repo": "Project Name/Infrastructure",
       "fileName": "default.json"
```

I had a look [at the source](https://github.com/renovatebot/renovate/blob/8bd77f747b3da59cc955b2fc4ba3301bcb680ea2/lib/modules/platform/azure/util.ts#L181) to find out how project name and repo name are looked up, and noticed the use of the Azure DevOps API.
So I've used the REST API to get all repositories and this is the part of the Infrastructure repo:
```json
{
    "id": "<Redacted>",
    "name": "Infrastructure",
    "url": "https://dev.azure.com/<Redacted>/<Redacted>/_apis/git/repositories/<Redacted>",
    "project": {
        "id": "<Redacted>",
        "name": "Project Name",
        "description": "<Redacted>",
        "url": "https://dev.azure.com/<Redacted>/_apis/projects/<Redacted>",
        "state": "wellFormed",
        "revision": 236,
        "visibility": "private",
        "lastUpdateTime": "2021-10-22T12:46:46.16Z"
    },
    "defaultBranch": "refs/heads/main",
    "size": 280323,
    "remoteUrl": "https://<Redacted>@dev.azure.com/<Redacted>/Project%20Name/_git/Infrastructure",
    "sshUrl": "git@ssh.dev.azure.com:v3/<Redacted>/Project%20Name/Infrastructure",
    "webUrl": "https://dev.azure.com/<Redacted>/Project%20Name/_git/Infrastructure",
    "isDisabled": false
}
```

So I'm 100% sure the repo name and project name match.
Does anyone have a clue why the repo would still be `null`?

**Reply by kipusoep** [reply]
Right, so the issue was that [the token that is used](https://github.com/jycouet/VSTSExtensions/blob/master/renovateMe/buildtask/src/index.ts#L21) didn't have permission to access the other repository.
What fixed it for us was to add:

```yaml
resources:
  repositories:
  - repository: <RepoName>
    type: git
    name: reponame

steps:
- checkout: reponame
```

This will trigger the permission request for the pipeline when you run it next time.
